query-planner-js/src/buildPlan.ts

```diff
@@ -969,20 +1140,46 @@ function sameMergeAt(m1: ResponsePath | undefined, m2: ResponsePath | undefined)
   return arrayEquals(m1, m2);
 }
 
+/**
+ * Used in `FetchDependencyGraph` to store, for a given group, information about one of its parent.
+ * Namely, this structure stores:
+ *  1. the actual parent group, and
+ *  2. the path of the group for which this is a "parent relation" into said parent (`group`). This information
+ *   is maintained for the case where we want/need to merge groups into each other. One can roughly think of
+ *   this as similar to a `mergeAt`, but that is relative to the start of `group`. 
+ */
+type ParentRelation = {
+  group: FetchGroup,
+  path?: OperationPath,
+}
+
+function concatPathsInParents(first: OperationPath | undefined, second: OperationPath | undefined): OperationPath | undefined  {
+  return first && second ? concatOperationPaths(first, second) : undefined;
+}
+
+function samePathsInParents(first: OperationPath | undefined, second: OperationPath | undefined): boolean  {
+  if (!first) {
+    return !second;
+  }
+  return !!second && sameOperationPaths(first, second);
+}
+
 class FetchDependencyGraph {
   private isReduced: boolean = false;
+  private isOptimized: boolean = false;
 
   private constructor(
     readonly subgraphSchemas: ReadonlyMap<string, Schema>,
     readonly federatedQueryGraph: QueryGraph,
     private readonly rootGroups: MapWithCachedArrays<string, FetchGroup>,
     private readonly groups: FetchGroup[],
-    private readonly adjacencies: number[][],
-    private readonly inEdges: number[][],
-    // For each groups, an optional path in its "unique" parent. If a group has more than one parent, then
-    // this will be undefined. Even if the group has a unique parent, it's not guaranteed to be set.
-    private readonly pathsInParents: (OperationPath | undefined)[]
-  ) {}
+    private readonly edges: {
+      outs: FetchGroup[],
+      ins: ParentRelation[],
+    }[],
+  ) {
+    assert(groups.length === edges.length, () => `Invalid arguments: ${groups.length} groups provided, but the edges array is only of size ${edges.length}`);
```

Comment:
There is a relationship here between the groups and the edges? Are you sure you don't want the type to be:

```
items: {
  group: FetchGroup,
  edge: {
    outs: FetchGroup[],
    ins: ParentRelation[],
  }
}[],
```

Reply:
Yeah, almost did that and worried it would make the diff for this issue even bigger for not that much added value. But agreed we can push the refactor a tad more and so went ahead. That said, instead of what's above, I moved the `outs` and `ins` directly into `FetchGroup` as I feel it's even a bit cleaner since:
1. the suggestion above would only help when iterating on all groups, which we almost never do, but when you have a group and want to know it's edges, you'd still have to index into `items`. Having the edges directly in the group avoids that.
2. almost all the methods on parents and children we were using were on `FetchGroup` anyway, so having the whole logic here save a bit of indirection of method we were having.